aptos-move/block-executor/src/executor.rs

```diff
@@ -1233,8 +1240,17 @@ where
             Version = Option<TxnIndex>,
         >,
     ) -> Result<(), PanicError> {
-        let (id, write_op) = write.unpack();
+        // Enforce read-before-write because storage (DB) relies on this assumption.
+        let _ = base_view.get_state_value(&key).map_err(|err| {
```

Comment:
we need some tests for this assumption 🤦‍♂️

Reply:
@msmouse Do you have a code pointer how I can test this?

Reply:
I think he adds a panic here https://github.com/aptos-labs/aptos-core/pull/15283 so we just need to test such some publish scenarios (with real db)

Reply:
many tests fail with the panic.. I need to debug today to see how to enforce the assumption better. 

none the less I think we need to trigger some module republishing in a smoke test

Reply:
@msmouse does it need to be a smoke test? we can have executor benchmark lib unit test with publishing with 3 lines of code

Reply:
it's hard to construct some complex scenario in smoke test because most of the time a block only has one txn (block produces too fast). it's easier to use executor + db alone to reproduce. I thought we already have module publish in smoke tests

Reply:
```

    #[test]
    fn test_module_publish_txns() {
        AptosVM::set_num_shards_once(1);
        AptosVM::set_concurrency_level_once(4);
        AptosVM::set_processed_transactions_detailed_counters();
        test_generic_benchmark::<AptosVMBlockExecutor>(
            Some(TransactionTypeArg::PublishPackage),
            true,
        );
    }

this doesn't call on them afterwards though
```

Reply:
we need to repeatedly publish the same module and the module must be being used so it resides in cache, is that simple to implement?

Reply:
Should be, let me see

Reply:
here is the republish and call mixed workload - #15292, and it fails.

you can use it to confirm/test this fix.

Reply:
my paranoid panic pr landed, btw

Reply:
Shouldn't we just remove this assumption or deal at the storage layer wherever it causes problems? Not just for modules, but in general it would be great if MoveVM/BlockSTM could avoid unnecessary reads - of course we could ignore for conflicts and fake for storage, like here, but that's uglier and error prone from here?

Reply:
I think we should really drain readsets captured by execution then? We have all the info, whatever DB is doing with the reads, we already know one layer above.

Reply:
yeah this is just for very short term mitigation. Going forward two things will happen:
1. the smt update will happen off the critical path, so we don't need to eager- read the proofs before the write set is formed. 
But 2. when we create the "hot tier" of state storage, we need to move state items between the hot tier and cold tier deterministically, that's where knowing precisely what gets read (for example, speculative reads that don't actually happen in the end need to be excluded) is REQUIRED. -- I'm gonna need to get the read set from the blockStm directly.

Reply:
we have deterministic `ReadWriteSummary` for each transaction, produced by BlockSTM - which we use for determining when block gas limit has been hit.

We can probably provide that/variant of that to storage. For storage to use it - would it be required to add it to WriteSet, or?

Reply:
I think making it part of the BlockEndInfo is proper.

Reply:
but then BlockEndInfo would be read set for the whole block, which might be large? but that might be a good place for it.

Reply:
yesh that's what I want. the trees like "bathcing effect"

Reply:
@msmouse that sounds great, one note is that we may have partial reads in the future, such as an existence check (can always be interpreted as a full read in lower layers, but in case it makes any difference)